File: 1/day1.py
from collections import Counter
def get_lists(file_path: str) -> tuple[list[int], list[int]]:
    l1, l2 = [], []
    with open(file_path) as f:
        for line in f:
            n1, n2 = list(map(int, line.split()))
            l1.append(n1)
            l2.append(n2)
    return l1, l2
    
def day1a(l1: list[int], l2: list[int]) -> int:
    l1.sort()
    l2.sort()
    return sum(abs(e1 - e2) for e1, e2 in zip(l1, l2))

def day1b(l1: list[int], l2: list[int]) -> int:
    """Gets similarity score between 2 lists."""
    hist2 = Counter(l2)
    return sum(e * hist2[e] for e in l1)

def main(file_path):
    l1, l2 = get_lists(file_path)
    # print(day1a(l1, l2))
    print(day1b(l1, l2))
    
if __name__ == '__main__':
    # main('./1/input0.txt')
    main('./1/input.txt')

File: 1/day1.js
const fs = require('fs');

function getLists(filePath) {
  const l1 = [];
  const l2 = [];
  const lines = fs.readFileSync(filePath, 'utf8').split('\n');

  for (const line of lines) {
    if (line.trim() === '') continue;
    const [n1, n2] = line.split(/\s+/).map(Number);
    l1.push(n1);
    l2.push(n2);
  }
  return [l1, l2];
}

function day1a(l1, l2) {
  l1.sort((a, b) => a - b);
  l2.sort((a, b) => a - b);
  return l1.reduce((acc, e1, i) => acc + Math.abs(e1 - l2[i]), 0);
}

function day1b(l1, l2) {
  const hist2 = l2.reduce((acc, num) => {
    acc[num] = (acc[num] || 0) + 1;
    return acc;
  }, {});
  return l1.reduce((acc, e) => acc + (hist2[e] || 0) * e, 0);
}

function main(filePath) {
  const [l1, l2] = getLists(filePath);
  console.log(day1a(l1, l2));
  console.log(day1b(l1, l2));
}

// Example usage
// main('./1/input0.txt');
// main('./1/input0.txt');
main('./1/input.txt');

/*
Input: input0.txt
Output: 
part1: 11
part2: 31

Inpput: input.txt
Output: 
part1: 2344935
part2: 27647262
*/